
###Concept
The idea of this program is to be able to call the lineage of a query MTB (Mycobacterium tuberculosis) strain quickly. You should be able to do this either starting from a genome assembly or from a vcf file.

###Requirements
* you need to install vrt_tools (check https://github.com/farhat-lab/vrt-tools)

####How to call the lineage starting from a complete genome (1 single scaffold)
1. We use Mummer to compare the two genomes
```
vrtTools-runMummer.py example/H37Rv_mod.fasta example/00-R0025.fasta mummerOut
```

2. We generate a .vrt file with the SNPs and INDELs
```
vrtTools-snps2vrt.py mummerOut.snps example/H37Rv_mod.fasta example/00-R0025.fasta ref_vs_00-R0025.vrt

```
3. We call the lineage
```bash
./bin/FastLineageCaller-assign2lineage.py example/db_snps.tsv ref_vs_00-R0025.vrt
--decision: lineage2,lineage2.2,lineage2.2.1,lineage2.2.1.1
```

The output shows the lineage and the sublineages

####How to call the lineage starting from a fasta file (multiple scaffolds, contigs)
1. we use Contiguator to generate a pseudocontig using the scaffolds/contigs

```
vrtTools-generatePseudocontig.py example/H37Rv_mod.fasta example/00-R0025_denovo.fasta
```

CONTIGuator will create the folders Map_<id_reference> and UnMappedContigs!


NOTE: CONTIGuator is very picky about the identifiers inside your fasta file. In order to avoid problems, please remove all dots present in the identifier before running vrtTools-generatePseudocontig.py. A command that can help you this this is:
```bash
cat H37Rv.fasta |sed -e 's/\./_/' > H37Rv_mod.fasta
```

2. We can now align the reference and the Pseudocontig with mummer
```bash
vrtTools-runMummer.py example/H37Rv_mod.fasta Map_NC.000962.3/PseudoContig.fsa out_mummer
```

3. We can now convert the .snps file generated by mummer into a .vrt file
```bash
vrtTools-snps2vrt.py out_mummer.snps example/H37Rv_mod.fasta Map_NC.000962.3/PseudoContig.fsa ref_vs_00-R0025.vrt
```

4. We call the lineage
```bash
./bin/FastLineageCaller-assign2lineage.py example/db_snps.tsv ref_vs_00-R0025.vrt
--decision: lineage2,lineage2.2,lineage2.2.1,lineage2.2.1.1
```

###How to call the lineage starting from a vcf file





###Pros and Cons


















####Callinng the lineage using a vcf file


~/lf61/check_strains_maha/12-strain-typing/bin/assign2lineage.py ~/lf61/check_strains_maha/12-strain-typing/data/db_snps.tsv out.vrt 
done > log_assignments.txt

cat log_assignments.txt |grep "^--" > tmp_assignments.txt

cat tmp_assignments.txt |tr -d "\n"|sed -e 's/.fasta//g'|sed -e 's/--decision://g'|sed -e 's/--strain://g'|sed -e 's/GCA/\nGCA/g' > final_table.tsv

#I generated an ods file -- it is easier to put some notes.
#I get the names of the strains
for i in `ls|grep fasta|sed -e 's/.fasta//'`;do printf ${i}"\t";head -1 ${i}.fasta|cut -d" " -f4-10|sed -e 's/, complete genome//';done


######################################
##I try to characterize M. africanum##
######################################
mkdir africanum
head -1 table_33894.tsv > table_complete_genomes_33894.tsv 
cat table_33894.tsv |grep "Complete Genome" >> table_complete_genomes_33894.tsv

#I downloaded the genomes by hand since there was a problem with the download script (ERROR 500: Internal Server Error)
gunzip *.fna.gz


#rename the files to fasta
for i in `ls|grep fna|sed -e 's/.fna//'`;do mv ${i}.fna ${i}.fasta;done

#I characterize the strains
for i in `ls|grep fasta`;do
echo "--strain: "${i}
~/lf61/mic_assemblies/18-pilot_cmp_pilon_denovo_00-R0025/bin/run_mummer.py ../../data/h37rv.fasta ${i} mummer
~/lf61/mic_assemblies/18-pilot_cmp_pilon_denovo_00-R0025/bin/snps2vrt.py mummer.snps ../../data/h37rv.fasta ${i} out.vrt
~/lf61/check_strains_maha/12-strain-typing/bin/assign2lineage.py ~/lf61/check_strains_maha/12-strain-typing/data/db_snps.tsv out.vrt 
done > log_assignments.txt

cat log_assignments.txt |grep "^--" > tmp_assignments.txt

cat tmp_assignments.txt |tr -d "\n"|sed -e 's/.fasta//g'|sed -e 's/--decision://g'|sed -e 's/--strain://g'|sed -e 's/GCA/\nGCA/g' > final_table.tsv


#I made my selection of strains. I am missing strains from lineages 5 and 7
#the final selection can be found on the directory selection_strains








