
### Concept
The idea of this program is to be able to call the lineage of a query MTB (Mycobacterium tuberculosis) strain quickly. You should be able to do this either starting from a genome assembly or from a vcf file.

### Requirements
* you need to install vrt_tools (check https://github.com/farhat-lab/vrt-tools)

#### How to call the lineage starting from a complete genome (1 single scaffold)
1. We use Mummer to compare the two genomes
```
vrtTools-runMummer.py example/H37Rv_mod.fasta example/00-R0025.fasta mummerOut
```

2. We generate a .vrt file with the SNPs and INDELs
```
vrtTools-snps2vrt.py mummerOut.snps example/H37Rv_mod.fasta example/00-R0025.fasta ref_vs_00-R0025.vrt

```
3. We call the lineage
```bash
FastLineageCaller-assign2lineage.py example/db_snps.tsv ref_vs_00-R0025.vrt
--decision: lineage2,lineage2.2,lineage2.2.1,lineage2.2.1.1
```

The output shows the lineage and the sublineages

#### How to call the lineage starting from a fasta file (multiple scaffolds, contigs)
1. we use Contiguator to generate a pseudocontig using the scaffolds/contigs

```
vrtTools-generatePseudocontig.py example/H37Rv_mod.fasta example/00-R0025_denovo.fasta
```

CONTIGuator will create the folders Map_<id_reference> and UnMappedContigs!


NOTE: CONTIGuator is very picky about the identifiers inside your fasta file. In order to avoid problems, please remove all dots present in the identifier before running vrtTools-generatePseudocontig.py. A command that can help you this this is:
```bash
cat H37Rv.fasta |sed -e 's/\./_/' > H37Rv_mod.fasta
```

2. We can now align the reference and the Pseudocontig with mummer
```bash
vrtTools-runMummer.py example/H37Rv_mod.fasta Map_NC.000962.3/PseudoContig.fsa out_mummer
```

3. We can now convert the .snps file generated by mummer into a .vrt file
```bash
vrtTools-snps2vrt.py out_mummer.snps example/H37Rv_mod.fasta Map_NC.000962.3/PseudoContig.fsa ref_vs_00-R0025.vrt
```

4. We call the lineage
```bash
FastLineageCaller-assign2lineage.py example/db_snps.tsv ref_vs_00-R0025.vrt
--decision: lineage2,lineage2.2,lineage2.2.1,lineage2.2.1.1
```

### How to call the lineage starting from a vcf file
1. We convert the vcf file to vrt
```bash
vrtTools-vcf2vrt.py example/00-R0025.vcf 00-R0025.vrt 1
```

2. We call the lineage
```
FastLineageCaller-assign2lineage.py example/db_snps.tsv 00-R0025.vrt
```








